### PR TITLE
JBPM-8155 - New preprocessor in custom queries

### PR DESCRIPTION
**Pull request by mcivantos-bbva**
Adding a new preprocessor in order to filter by potential owners. The preprocessor creates 2 types of filter:

If the custom query hasn't got any query-params related to the potential owners, the preprocessor adds a new column filter by the logged user and his groups.
If the custom query has a query-params related to a list of potential owners, the logged user won't be take into account in the filter, but the new preprocessor will add a new column filter by the list of potOwners included in the query-params and their groups.

### CODE REVIEW

**Comment by kiereleaseuser3** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessorTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        dataSetLookup.addOperation(filter);
+        
+        userAndGroupsTasksPreprocessor.preprocess(dataSetLookup);
+
+        assertEquals("(POTOWNER in role1, role2, " + potOwner + ")",
+                     dataSetLookup.getFirstFilterOp().getColumnFilterList().get(0).toString());
+    }
+    
+    @Test
+    public void testPotOwnerFilterNoIdentityProviderAndNullGoups() {
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessorTest.java]: typo Goups -> Groups

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+public class UserAndGroupsTasksPreprocessor extends UserTasksPreprocessor{
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(UserAndGroupsTasksPreprocessor.class);
+
+    private IdentityProvider identityProvider;
+
+    private UserGroupCallback userGroupCallback;
+    
+    private String columnId;
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Please fix the indentation in all files so it matches the code style in the rest of the repository.

Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+        ColumnFilter columnFilter;
+        List<Comparable> orgEntities = new ArrayList<>();
+        
+
+        boolean potOwnerFilter = false;
+
+        if (lookup.getFirstFilterOp() != null) {
+        	List<String> potOwners = new ArrayList<String>();
+        	List<ColumnFilter> columFilters = lookup.getFirstFilterOp().getColumnFilterList();
```
[the remark this answers is not in the document]
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        		if (column.getColumnId().toUpperCase().equals(columnId)) {
+        			potOwners.addAll(column.getParameters());
+        			
+        			for (String potOwner : potOwners) {
+        				orgEntities.addAll(Optional.ofNullable(userGroupCallback.getGroupsForUser(potOwner)).orElse(new ArrayList<>()));
+        				orgEntities.add(potOwner);
+        			}
+        			potOwnerFilter = true;
+        			int index = lookup.getFirstFilterOp().getColumnFilterIdx(column).intValue();
+            		lookup.getFirstFilterOp().getColumnFilterList().remove(index);
```
[the remark this answers is not in the document]
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: @MarianMacik I've followed your suggestions and I changed the code and I've also added a new test in order to check an empty list of potOwners.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        List<Comparable> orgEntities = new ArrayList<>();
+        
+
+        boolean potOwnerFilter = false;
+
+        if (lookup.getFirstFilterOp() != null) {
+        	List<String> potOwners = new ArrayList<String>();
+        	List<ColumnFilter> columFilters = lookup.getFirstFilterOp().getColumnFilterList();
+        	Iterator<ColumnFilter> it= columFilters.iterator();
+        	while ((potOwners.size() <= 0) && it.hasNext()) {
```
[the remark this answers is not in the document]
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        			potOwnerFilter = true;
+        			int index = lookup.getFirstFilterOp().getColumnFilterIdx(column).intValue();
+            		lookup.getFirstFilterOp().getColumnFilterList().remove(index);
+        		}
+        		
+        	}
+        	
+        	if (!potOwnerFilter) {
+        		orgEntities.addAll(Optional.ofNullable(userGroupCallback.getGroupsForUser(identityProvider.getName())).orElse(new ArrayList<>()));
+        		orgEntities.add(identityProvider.getName());
```
[the remark this answers is not in the document]
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessorTest.java (hunk 1 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessorTest.java]: Fixed

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java (hunk 2 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done too.

**Comment by MarianMacik** [comment]
ok to test

**Comment by MarianMacik** [comment]
jenkins retest this

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            Iterator<ColumnFilter> it= columnFilters.iterator();
+            while (orgEntities.isEmpty() && it.hasNext()) {
+                CoreFunctionFilter column = (CoreFunctionFilter)it.next();
+
+                if (column.getColumnId().toUpperCase().equals(columnId)) {
+                    potOwners.addAll(column.getParameters());
+
+                    for (String potOwner : potOwners) {
+                        orgEntities.addAll(Optional.ofNullable(userGroupCallback.getGroupsForUser(potOwner)).orElse(new ArrayList<>()));
+                        orgEntities.add(potOwner);
```
Comment (mswiderski) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: wouldn't it make sense to use `AddUserAndGroupsFromIdentityProvider` method here too? Just parameterise the userid...

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            filter.addFilterColumn(columnFilter);
+            lookup.addOperation(filter);
+        }
+
+        LOGGER.debug("Adding column filter: {}", columnFilter);
+
+        super.preprocess(lookup);
+    }
+
+    private void AddUserAndGroupsFromIdentityProvider(List<Comparable> orgEntities) {
```
Comment (mswiderski) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: please use lower case as first letter of the method name

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java (hunk 7 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java (hunk 8 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/QueryServiceImpl.java [hunk 9]
```diff
@@ -206,6 +208,8 @@ public void replaceQuery(QueryDefinition queryDefinition) {
                     dataSetDefRegistry.registerPreprocessor(sqlDef.getUUID(), new BusinessAdminTasksPreprocessor(identityProvider, userGroupCallback, metadata));
                 } else if (queryDefinition.getTarget().equals(Target.FILTERED_PO_TASK)) {
                     dataSetDefRegistry.registerPreprocessor(sqlDef.getUUID(), new PotOwnerTasksPreprocessor(identityProvider, userGroupCallback, metadata));
+                }else if (queryDefinition.getTarget().equals(Target.USER_GROUPS_TASK)) {
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/QueryServiceImpl.java]: Please correct the formatting here.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import org.dashbuilder.dataset.DataSetMetadata;
+import org.dashbuilder.dataset.filter.ColumnFilter;
+import org.dashbuilder.dataset.filter.CoreFunctionFilter;
+import org.dashbuilder.dataset.filter.DataSetFilter;
+import org.kie.api.task.UserGroupCallback;
+import org.kie.internal.identity.IdentityProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UserAndGroupsTasksPreprocessor extends UserTasksPreprocessor{
```
Comment (MarianMacik) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Again, formatting here, should be `UserTasksPreprocessor {`

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java [hunk 11]
```diff
@@ -33,7 +33,8 @@
         FILTERED_PROCESS,
         FILTERED_BA_TASK,
         FILTERED_PO_TASK,
-        CUSTOM;
+        CUSTOM,
+        USER_GROUPS_TASK;
```
Comment (MarianMacik) [inline jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java]: I would leave CUSTOM as the last one, by looking at the former PRs this was the convention before.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/QueryServiceImpl.java (hunk 9 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/QueryServiceImpl.java]: It's already fixed.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java (hunk 10 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java]: Done

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java (hunk 11 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java]: Done

Comment (MarianMacik) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java]: Thanks, Maria.

Comment (mcivantos-bbva) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/query/model/QueryDefinition.java]: My pleasure :-)

**Comment by MarianMacik** [comment]
@mcivantos-bbva I believe we need to add an ignore for the revapi...

**Comment by mcivantos-bbva** [comment]
@MarianMacik  adding ignore for the revapi and rebase done.

File: jbpm-services/jbpm-services-api/src/build/revapi-config.json [hunk 12]
```diff
@@ -45,7 +45,28 @@
...
+        },
+         {
+         "code": "java.method.addedToInterface",
+         "new": "method java.util.List<org.kie.internal.runtime.error.ExecutionError> org.jbpm.services.api.admin.ProcessInstanceAdminService::getErrorsByDeploymentId(java.lang.String, boolean, org.kie.api.runtime.query.QueryContext)",
+         "package": "org.jbpm.services.api.admin",
+         "classSimpleName": "ProcessInstanceAdminService",
+         "methodName": "getErrorsByDeploymentId",
+         "elementKind": "method",
+         "justification": "Added support for filtering errors by deployment id"
+         },
```
Comment (MarianMacik) [inline jbpm-services/jbpm-services-api/src/build/revapi-config.json]: Is this needed as well? You didn't do any changes related to ProcessInstanceAdminService...

File: jbpm-services/jbpm-services-api/pom.xml [hunk 13]
```diff
@@ -6,6 +6,7 @@
     <version>7.18.0-SNAPSHOT</version>
   </parent>
   <artifactId>jbpm-services-api</artifactId>
+  <version>7.17.0-SNAPSHOT</version>
```
Comment (MarianMacik) [inline jbpm-services/jbpm-services-api/pom.xml]: I believe that the build failed with a compilation error because this version was changed to the older one. It shouldn't be here at all.

Comment (mcivantos-bbva) [reply jbpm-services/jbpm-services-api/pom.xml]: Sorry! It was an error due to the rebase. Fixed.

File: jbpm-services/jbpm-services-api/src/build/revapi-config.json (hunk 12 above)
Comment (mcivantos-bbva) [reply jbpm-services/jbpm-services-api/src/build/revapi-config.json]: Done
